### ISSUE TITLE
gpu_state_isolation: fails on coder-0.5b, and its own message truncates both outputs to an identical prefix so it never shows the difference

### ISSUE THREAD

**Issue by noahgift** [paiml/aprender#3824]
Found while diagnosing #3804, and deliberately **not** folded into it: a second real defect is not evidence for the first.

## Measured

`apr 0.69.0 (a9502d992)`, RTX 4090 sm_89, `qwen2.5-coder-0.5b-instruct-q4_k_m.gguf`:

```
apr qa .../qwen2.5-coder-0.5b-instruct-q4_k_m.gguf --skip-throughput --skip-ollama --json
```

```
gpu_state_isolation FAIL —
State leak: prompt A produced different output on retry.
  First: '<|im_start|>user\nWhat is 2+2?<|im_end|>\n<|im_start'
  Retry: '<|im_start|>user\nWhat is 2+2?<|im_end|>\n<|im_start'
```

Two separate problems here.

## 1. The gate's message cannot show what it claims

**The `First:` and `Retry:` strings printed above are identical.** Both are truncated to the same ~50-character prefix, so whatever diverged is past the cut. The failure message therefore proves nothing to a reader and gives no lead — you cannot tell from the output whether the divergence was at token 12 or token 120, or how large it was.

`crates/apr-cli/src/commands/gpu_isolation_result.rs:273` (`GpuIsolationResult::StateLeak(first, retry)`) should report the **first differing token index** and a window around it, not two equal prefixes. A gate that fails without showing the difference is a gate you cannot act on.

## 2. The underlying non-determinism is NOT in the diverging forward path

Worth recording precisely, because it is easy to assume otherwise and doing so would misdirect a bisect (it nearly misdirected mine).

The GPU forward that produces #3804's wrong logits is **perfectly deterministic**. Measured with a per-position cosine dump over **10 fresh processes**, `--revalidate` so no receipt short-circuits the guard: byte-identical to 6 decimals at all 16 probe positions in every run —

```
pos 0: cosine 0.099921  cpu_argmax 144328  gpu_argmax 80984
pos 1: cosine 0.415256  cpu_argmax  73562  gpu_argmax 20840
...
pos15: cosine 0.867281  cpu_argmax    488  gpu_argmax    15
```

So whatever `gpu_state_isolation` is detecting lives in the **repeat-generation path inside the qa gate** (KV-cache or workspace state not reset between the first and retry generation), not in the forward that #3804 is about. Those are different defects and the fix for one will not fix the other.

## Ask

1. Make `StateLeak` report the first differing token index and a window around it, so the message carries the evidence for its own claim.
2. Determine what state is not reset between the two generations in the isolation gate.

## Note on scope

This model also fails `golden_output` for an unrelated, now-identified reason (#3804: FP8 E4M3 batched prefill). Reproducing this row on a model that passes `golden_output` would isolate it cleanly.

Refs #3804.

**Comment by noahgift** [comment]
milestone → 0.71.0 per docs/roadmaps/milestone-triage-2026-09-23.md (#4024), operator approval 2026-09-23 (relayed by aprender-cf). Reason: model/backend correctness or the certified matrix: Don't Leave Behind (#3994)
